staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,619 @@
+			qs.GetRequestsExecuting() == 0 {
+			qs.counter.Add(1)
+			go qs.emptyHandler.HandleEmpty()
+		}
+	}
+}
+
+// dequeue dequeues a packet from the queueSetImpl
+func (qs *queueSetImpl) dequeue() (*fq.Request, bool) {
+
+	queue := qs.selectQueue()
+
+	if queue == nil {
+		return nil, false
+	}
+	packet, ok := queue.Dequeue()
+
+	if ok {
+		// When a request is dequeued for service -> qs.VirStart += G
+		queue.VirStart += qs.estimatedServiceTime
+
+		packet.StartTime = qs.clk.Now()
+
+		// request dequeued, service has started
+		queue.RequestsExecuting++
+		metrics.UpdateFlowControlRequestsExecuting(qs.name, queue.RequestsExecuting)
+	} else {
+		// TODO(aaron-prindle) verify this statement is needed...
+		return nil, false
+	}
+	qs.numRequestsEnqueued--
+	return packet, ok
+}
+
+// Dequeue dequeues a packet from the queueSetImpl
+func (qs *queueSetImpl) Dequeue() (*fq.Request, bool) {
+	qs.lockAndSyncTime()
+	defer qs.lock.Unlock()
+	return qs.dequeue()
+}
+
+// isEmpty is a convenience method that returns 'true' when all of the queues
+// in an queueSetImpl have no packets (and is "empty")
+func (qs *queueSetImpl) isEmpty() bool {
+	return qs.numRequestsEnqueued == 0
+}
+
+// DequeueWithChannelAsMuchAsPossible runs a loop, as long as there
+// are non-empty queues and the number currently executing is less than the
+// assured concurrency value.  The body of the loop uses the fair queuing
+// technique to pick a queue, dequeue the request at the head of that
+// queue, increment the count of the number executing, and send `{true,
+// handleCompletion(that dequeued request)}` to the request's channel.
+func (qs *queueSetImpl) DequeueWithChannelAsMuchAsPossible() {
+	for !qs.isEmpty() && qs.GetRequestsExecuting() < qs.concurrencyLimit {
+		_, ok := qs.dequeueWithChannel()
+		// TODO(aaron-prindle) verify checking ok makes senes
+		if !ok {
+			break
+		}
+	}
+}
+
+// dequeueWithChannel is convenience method for dequeueing packets that
+// require a message to be sent through the packets channel
+// this is a required pattern for the QueueSet the queueSetImpl supports
+func (qs *queueSetImpl) dequeueWithChannel() (*fq.Request, bool) {
+	pkt, ok := qs.dequeue()
+	if !ok {
+		return nil, false
+	}
+	qs.counter.Add(1)
+	pkt.DequeueChannel <- true
+	return pkt, ok
+}
+
+func (qs *queueSetImpl) roundrobinqueue() int {
+	// TODO(aaron-prindle) verify this is modified on quiesce...
+	qs.robinIdx = (qs.robinIdx + 1) % len(qs.queues)
+	return qs.robinIdx
+}
+
+// selectQueue selects the minimum virtualfinish time from the set of queues
+// the starting queue is selected via roundrobin
+// TODO(aaron-prindle) verify that the roundrobin usage is correct
+// unsure if the code currently prioritizes the correct queues for ties
+func (qs *queueSetImpl) selectQueue() *fq.Queue {
+	minvirfinish := math.Inf(1)
+	var minqueue *fq.Queue
+	var minidx int
+	for range qs.queues {
+		// TODO(aaron-prindle) how should this work with queue deletion?
```

Comment:
With queue deletion `qs.robinIdx` should be updated to maintain the invariant that `(qs.robinIdx+1) % numQueues` is the index of the queue after the one last selected.  See other comments about making this true.

Reply:
Removed this TODO and use the suggested method stated elsewhere